### PR TITLE
baremetal: send full ignition to masters

### PR DESCRIPTION
**Pull request by hardys**
This restores the work which was previously done via #3276 but then reverted via #3589 due to breaking users who customized the pointer ignition config in IPI deployments.

A solution to that has been proposed via https://github.com/openshift/installer/pull/4413 which this PR depends on, so when that lands this work can safely be restored, see openshift/enhancements#540 for more details.

Note that some additional changes beyond the initial implementation were required due to the MCO now supporting multiple
ignition versions, thus this PR depends on https://github.com/openshift-metal3/terraform-provider-ironic/pull/46 which will need to land and be vendored here before the PR merges.

This replaces https://github.com/openshift/installer/pull/4359

### CODE REVIEW

File: pkg/tfvars/baremetal/baremetal.go
Comment (stbenjam) [inline pkg/tfvars/baremetal/baremetal.go:172]: These two slice accesses should check to make sure those elements exist first and return an error if not.

File: pkg/asset/ignition/bootstrap/bootstrap.go
Comment (stbenjam) [inline pkg/asset/ignition/bootstrap/bootstrap.go:87]: Nit, `Check` sounds passive to me, although these two assets are generating machine configs.  Maybe `Convert` or something more active?

**Review by stbenjam — CHANGES_REQUESTED** [review]
A couple minor comments on a first pass, not super familiar with the RawExt ignition stuff. Would be good to get MCO folks to review.

Note, you'll also need to update the [dependency graph](https://github.com/openshift/installer/blob/master/docs/design/assetgeneration.md#dependency-graph).

File: pkg/asset/ignition/bootstrap/bootstrap.go
Comment (kirankt) [reply pkg/asset/ignition/bootstrap/bootstrap.go:87]: +1. The names were placeholders (naming is hard!!), so something like "convert" makes sense.

Comment (hardys) [reply pkg/asset/ignition/bootstrap/bootstrap.go:87]: Yeah @staebler suggested `MasterIgnitionCustomizations` on https://github.com/openshift/installer/pull/4413 which works for me, unless there are any objections I'll update 4413 with that name and rebase this accordingly.

File: pkg/tfvars/baremetal/baremetal.go
Comment (hardys) [reply pkg/tfvars/baremetal/baremetal.go]: Added a check to ensure they aren't zero length

**Comment by hardys** [comment]
> Note, you'll also need to update the dependency graph.

Ah I missed that in 4413, I'll push a follow-up, thanks!

File: pkg/asset/cluster/tfvars.go
Comment (staebler) [inline pkg/asset/cluster/tfvars.go:96]: Why is this being added as a dependency?

File: data/data/baremetal/masters/variables.tf
Comment (staebler) [inline data/data/baremetal/masters/variables.tf:32]: Include "master" in the names of these variables to make it clear that they are for the masters and not the bootstrap.

File: pkg/tfvars/baremetal/baremetal.go
Comment (staebler) [inline pkg/tfvars/baremetal/baremetal.go:43]:
```suggestion
func TFVars(libvirtURI, bootstrapProvisioningIP, bootstrapOSImage, externalBridge, externalMAC, provisioningBridge, provisioningMAC string, platformHosts []*baremetal.Host, image, ironicUsername, ironicPassword, masterIgnition string) ([]byte, error) {
```

File: pkg/asset/cluster/tfvars.go
Comment (hardys) [reply pkg/asset/cluster/tfvars.go:96]: Good point @kirankt can you remember if/why this is required?

Comment (hardys) [reply pkg/asset/cluster/tfvars.go:96]: I suspect it's not required since the masterIgn assed already has this dep ref https://github.com/openshift/installer/blob/master/pkg/asset/ignition/machine/master.go#L32

I'll re-test without it and remove provided everything still works :)

Comment (kirankt) [reply pkg/asset/cluster/tfvars.go:96]: We don't need this. It is a mistake.

File: data/data/baremetal/masters/variables.tf
Comment (kirankt) [reply data/data/baremetal/masters/variables.tf:32]: This one is already nested under masters directory, so can't we infer that these variables belong to the master? Besides, I think this change will impact the TF provider code as well.

Comment (staebler) [reply data/data/baremetal/masters/variables.tf:32]: Sorry, I clicked the wrong file. I mean the ones [here](https://github.com/openshift/installer/pull/4427/files#diff-328dc01a4376155918bed799b815ab5731f7dec17ab3e8ea83a40b581fc6e018R56-R69)

Comment (staebler) [reply data/data/baremetal/masters/variables.tf:32]: And [here](https://github.com/openshift/installer/pull/4427/files#diff-3b0c0219b1206253af5d430d464ea73c976c339a055510a0d9456d4dfb0419afR30-R32)

File: pkg/asset/cluster/tfvars.go
Comment (hardys) [reply pkg/asset/cluster/tfvars.go]: Now removed.

File: data/data/baremetal/masters/variables.tf
Comment (hardys) [reply data/data/baremetal/masters/variables.tf]: I updated the terraform variable names (not in the TF provider only the installer tf files), but I missed the code variable names (that's not exposed to the user but I'll update for consistency)

Comment (hardys) [reply data/data/baremetal/masters/variables.tf]: Now done

File: data/data/baremetal/masters/variables.tf
Comment (staebler) [inline data/data/baremetal/masters/variables.tf:32]: Sorry for the confusion around the naming. These didn't **need** to change, as you correctly stated that there is already context in the path. You can leave it as is or change it back: Your call.

File: pkg/tfvars/baremetal/baremetal.go
Comment (staebler) [reply pkg/tfvars/baremetal/baremetal.go:43]: This change was not made, but the comment was resolved. Is that an oversight? Or are you comfortable with the `ignition` name for the variable?

File: data/data/baremetal/masters/variables.tf
Comment (hardys) [reply data/data/baremetal/masters/variables.tf:32]: @staebler since CI passed I'm fine to leave it with the master prefix - it also makes it slightly more explicit when debugging and looking at the tfvars directly I guess

**Comment by stbenjam** [comment]
/test e2e-metal-ipi-virtualmedia

Just want to test the disabled/virtualmedia code path since it's so fragile, but the changes here look good to me!

**Comment by stbenjam** [comment]
@staebler Is there anything left from your side to get this approved? We'll need you to do it :)

File: pkg/tfvars/baremetal/baremetal.go
Comment (staebler) [inline pkg/tfvars/baremetal/baremetal.go:43]: I was still hoping to get this changed from `ignition` to `masterIgnition`, but I won't hold up the PR for it.

**Review by staebler — COMMENTED** [review]
/approve

**Review by kirankt — APPROVED** [review]

File: pkg/tfvars/baremetal/baremetal.go
Comment (hardys) [reply pkg/tfvars/baremetal/baremetal.go:43]: Thanks, I can push a follow-up PR to make that consistent

**Pull request merged**
